book/src/qobject/qobject_struct.md

```diff
@@ -5,42 +5,134 @@ SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
+
+## Requirements
+- Like most other CXX-Qt macros, it can only be used from within a [`#[cxx_qt::bridge]`](./bridge-macro.md).
+- The `#[cxx_qt::qobject]` macro must be placed on a Rust struct.
+- The struct must `impl Default`, so that it can be constructed as part of a QObject.
+
+## Effects
+Adding the macro to a Rust struct `MyObject` has a few effects.
+
+However, first it's important to mention that this macro **does not modify the contents of the struct in any way!**
+This means you can always rely on the struct to behave like any other Rust struct.
+You can use it in normal Rust code, without interacting with Qt in any way.
+
+The macro does multiple other things for you though:
+- Generate a C++ QObject subclass that wraps the `MyObject` Rust struct.
+- Expose the generated QObject subclass as `qobject::MyObject`. (For details see the [`qobject::T`](./cpp_object.md) page).
+- Generate getters/setters for all fields.
+- Generate `Q_PROPERTY`s for all fields that are marked as `#[qproperty]`.
+- Generate signals if paired with a [`#[cxx_qt::qsignals]` enum](./signals_enum.md).
+
+### `base` attribute
+Use the `base` attribute to specify a C++ class that the C++ QObject will inherit from.
+The base class must be a QObject!
+
+``` rust,ignore,noplayground
+{{#include ../../../examples/qml_features/rust/src/custom_base_class.rs:book_qobject_base}}
+      // ...
+    }
+```
+
+Also make sure to use the CXX `include!` macro to include the appropriate C++ header for the base class.
+``` rust
+{{#include ../../../examples/qml_features/rust/src/custom_base_class.rs:book_base_include}}
+    }
+```
 
-The `#[cxx_qt::qobject]` marked struct allows you to define the following items
+[Full Example](https://github.com/KDAB/cxx-qt/blob/main/examples/qml_features/rust/src/custom_base_class.rs)
 
-  * The name of the C++ class for the QObject
-  * Invokable methods that are exposed to Qt
-  * Private methods and fields for struct to use (eg this is useful for storing the channels for [threading](../concepts/threading.md))
-  * Mutate C++ state with [`CppObj`](./cpp_object.md)
-  * Use [CxxQtThread<T>](./cxxqtthread.md) for multi threaded workflows
+## Properties
+
+Fields within the `#[cxx_qt::qobject]` marked struct can be tagged with `#[qproperty]`, this declares which fields will be exposed as `Q_PROPERTY`s on the QObject.
 
 ```rust,ignore,noplayground
-{{#include ../../../examples/qml_features/rust/src/rust_obj_invokables.rs:book_macro_code}}
+{{#include ../../../examples/qml_features/rust/src/properties.rs:book_macro_code}}
+```
+
+Any type that CXX supports may be marked as a `#[qproperty]`.
+See the [Types page](../concepts/types.md) for a list of supported types.
+
+For every `#[qproperty]`, CXX-Qt will generate setters and getters, as well as a "changed" signal.
+On The C++ side:
+- setter: `set<Property>` - getter: `<Property>`
+On the Rust side:
+- setter: `set_<Property>` - getter: `<Property>`
+where `<Property>` is the name of the property
```

Comment:
The mdBook formatting isn't working well here. Maybe use a 2x2 table (language by setter/getter)?

Reply:
I've updated this to split it into separate paragraphs, which is hopefully good enough now :)